src/dataflow/src/sink/tail.rs

```diff
@@ -66,6 +70,29 @@ pub fn tail<G>(
             // though it is slower because it will produce deterministic results since the
             // cursor will always produce rows in the same order.
             results.sort_by_key(|(time, _)| *time);
+
+            if connector.emit_timestamps && !batches.is_empty() {
+                // The user has request timestamp done messages. We've already sorted the
+                // results by timestamp so we can duplicate the last row's timestamp which is
+                // somewhat easier than using a batch.upper() antichain. This timestamp won't
+                // be reused any other batch because they are disjoint and strictly increasing.
+                let (time, _) = results.last().unwrap().clone();
+                packer.push(Datum::Decimal(Significand::new(i128::from(time))));
+                packer.push(Datum::True);
+                // Fill in all table columns with NULL. Determine the number of columns
+                // by iterating through a row and skipping the first two that we add by
+                // hand. Remember this value and only iterate once because the iteration
+                // involves decoding the row datum bytes which is slow.
+                if cols == 0 {
```

Comment:
I think if you were sinking a view with 2 columns this would do a lot more row decoding -- might be better to use an option for cols?

Reply:
The `skip(2)` skips the `timestamp` and `done` columns. It should always be 1 (for `diff`) + (number of cols in the thing being tailed).

Reply:
Added a comment to clarify.

Reply:
ah I see. I had glossed over the part where we add extra fields to the data rows. got it!